### DISCUSSION TITLE
How to run a script when viewport changes?

### DISCUSSION THREAD

**Discussion by vdegenne** [WayfireWM/wayfire#2470] [Q&A] [answered]
I have a viewport with my IDE open and on the viewport on the right I have the browser, when I change a file and save the browser updates.
I made a script to save the file in nvim, now I just need to execute it when the viewport change so I wouldn't have to press ctrl+s to save every time I want to see the changes. 

What would be the best way of detecting viewport changes in wayfire?

The only thing I can think of right now is using IPC but I'm trying to find out if there is an easier way to do that.

Thanks for your help

**Comment by soreau** [comment, marked as answer]
IPC is probably the best way to go, especially if you're using latest wayfire. You can use `sock.watch["wset-workspace-changed"]` to receive [this data set](https://github.com/WayfireWM/wayfire/blob/master/plugins/ipc-rules/ipc-events.hpp#L330). For this to work, you will need to enable ipc and ipc-rules plugins, and install pywayfire with python pip. The other way is to write a wayfire plugin which calls `wf::get_core().run(cmd);` when it gets the `wf::workspace_changed_signal` signal internally.

**Reply by vdegenne** [reply]
Thanks! Why would you say the second method you suggest is not the "best"? I'm not sure what a wayfire plugin is but for me "best" is if I can write less.

**Reply by soreau** [reply]
I would say that writing a plugin to just run a command on a signal is overkill when there is the simplistic option of wiring an ipc python script in a handful of lines.

**Reply by vdegenne** [reply]
for anyone interested:

```python
from wayfire import WayfireSocket
import subprocess

events_sock = WayfireSocket()
events_sock.watch(["wset-workspace-changed"])

while True:
    events_sock.read_message()
    subprocess.run(['$HOME/.config/tmux/tmux-save-nvim.sh'], shell=True)

```

**Reply by soreau** [reply]
For what it's worth, you don't need to get the environment variable `WAYFIRE_SOCKET`, you can just call ws.WayfireSocket() without any argument.

**Reply by vdegenne** [reply]
I tried but I got an error so this parameter seems compulsory. But I removed the condition because this watch only covers one event.

**Reply by soreau** [reply]
See [here](https://github.com/WayfireWM/pywayfire/blob/main/scripts/ipc-rules-demo.py#L12).

**Reply by vdegenne** [reply]
Hmm, I tried that but it says "No module named 'wayfire'"

**Reply by soreau** [reply]
As I said, you will need to install pywayfire with python pip.

**Reply by vdegenne** [reply]
Nice thanks! Btw I don't think that's a good solution because when i go on the right and back on left it will execute the script twice, eventually leading to glitches.
One better solution would be to detect when tmux is losing focus, I don't know if IPC can go that deep.
I could also always stick to a viewport for my IDE and use the data from the message to avoid saving when my IDE gains focus again, but sometimes I have multiple IDE open on different viewports so it can get a little tricky.

**Reply by soreau** [reply]
You should be able to check ["new_workspace"](https://github.com/WayfireWM/wayfire/blob/master/plugins/ipc-rules/ipc-events.hpp#L336) auxiliary data, and make sure it's the workspace you want before running the command.

**Reply by soreau** [reply]
To check for view focus lost, you can watch for "view-focused" event, and record the view app-id. If the last app-id was tmux and the current is not, run the command.

**Reply by vdegenne** [reply]
Awesome thanks a lot. Here's the final script: 

```python
from wayfire import WayfireSocket
import subprocess

TARGET_WINDOW = "Alacritty"

events_sock = WayfireSocket()
events_sock.watch(["view-focused"])

last_window = None

while True:
    msg = events_sock.read_message()

    if msg["view"] is not None:
        current_window = msg["view"]["app-id"]

        if last_window == TARGET_WINDOW:
            subprocess.run(
                ['$HOME/.config/tmux/tmux-save-nvim.sh'], shell=True)

        last_window = current_window
```
